feat: adjust GitHub Actions documentation for new mittwald/zerodeploy-action (#1124)
mittwald/developer-portal: docs/guides/deployment/container-actions.mdx

> Using GitHub Actions you can build and publish your container image right from your source code repository, and then immediately deploy your application on the mittwald cloud platform. The official **`mittwald/deploy-container-action`** takes care of the API calls, applies your stack definition, and (if necessary) re-creates the affected services.

Reviewer: This sentence ("The official **`mittwald/deploy-container-action`** takes care of ...") still implies there is only *one* action available for use.

## Prerequisites {#prerequisites}

- **mStudio API token**; see the docs on [obtaining an API token](/docs/v2/api/intro#obtaining-an-api-token)
- **GitHub Actions runner**; any runner with Docker support, e.g. `ubuntu-latest`

## Which action should you use? {#which-action-should-you-use}

You now have two different deployment styles:

| Use case | Recommended action | Why |
| --- | --- | --- |
| You want to deploy quickly from source code without creating Dockerfiles or stack files first | **`mittwald/zerodeploy-action`** | Uses Railpack to infer build steps and deploys directly to mittwald container hosting with near-zero config. |
| You already have a Dockerfile and/or explicit stack definition (`stack.yaml`) and want full control over services, ports, and rollout behavior | **`mittwald/deploy-container-action`** | Best when you want an explicit, infrastructure-focused deployment model. |

If you want the fastest path from repository to running service, use `zerodeploy-action`. If your team already maintains container build and runtime definitions, stay with `deploy-container-action`.

> ## Deploy with `mittwald/zerodeploy-action` (Railpack) {#deploy-with-zerodeploy-action-railpack}

Reviewer: May be a matter of preference, but I would prefer not leaking too much implementation details (using Railpack, for example) to the end user. In the end, this implementation detail is of little consequence for the end user and may be distracting.

```suggestion
## Deploy with `mittwald/zerodeploy-action` {#deploy-with-zerodeploy-action}
```

> The [`mittwald/zerodeploy-action`](https://github.com/mittwald/zerodeploy-action) is designed for repositories that do not have a Docker setup yet. It installs the mittwald CLI, installs Railpack, builds from your repository, and deploys to your target project.

Reviewer: I would not go too much in to detail as to _how_ the action actually does what it does (see the other comment about leaking implementation details).

```suggestion
The [`mittwald/zerodeploy-action`](https://github.com/mittwald/zerodeploy-action) is designed for repositories that do not have a Docker setup yet. It automatically determines the best way to build a Docker image from your code, and deploys that to your target project, using a dynamically provisioned container registry.
```

### Required inputs {#required-inputs-zerodeploy}

- **`MITTWALD_API_TOKEN`** secret; API token from mStudio
- **Project ID** (for example `p-XXXXXX`)

### Example workflow {#example-workflow-zerodeploy}

Create `.github/workflows/zerodeploy.yml`:

```yaml
name: Deploy to mittwald Container Hosting

on:
  workflow_dispatch:

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

>       - name: Create .env for deployment

Reviewer: I'm also wondering, if we can (for a later change) find a more convenient solution for injecting environment variables... maybe add an additional parameter to the zerodeploy-action that writes the `.env` file implicitly?

Author: Yeah that part smells a little. Experimented a little with implicit Github environment serialization, maybe that's a ( the ) way to go in the future.

        env:

>           APP_SECRET: ${{ secrets.APP_SECRET }}
>         run: |
>           {
>             echo "APP_ENV=production"
>             echo "APP_SECRET=$APP_SECRET"
>           } > .env

Reviewer: Can be shortened:

```suggestion
      - name: Create .env for deployment
        run: |
          {
            echo "APP_ENV=production"
            echo "APP_SECRET=${{ secrets.APP_SECRET }}"
          } > .env
```

Using this workflow, on every push of a new Git tag (or on any push to the `main` branch, depending on which configuration you choose), GitHub will build a new image, push it to GitHub Container Registry and instruct mStudio to run it.

> ## Which action should you use? {#which-action-should-you-use}

Reviewer: Maybe move this section to the very top?

Author: Good one. That will make it easier to read and find the information needed :+1:

You now have two different deployment styles:

| Use case | Recommended action | Why |
| --- | --- | --- |
| You already have a Dockerfile and/or explicit stack definition (`stack.yaml`) and want full control over services, ports, and rollout behavior | **`mittwald/deploy-container-action`** | Best when you want an explicit, infrastructure-focused deployment model. |
| You want to deploy quickly from source code without creating Dockerfiles or stack files first | **`mittwald/zerodeploy-action`** | Uses Railpack to infer build steps and deploys directly to mittwald container hosting with near-zero config. |

If your team already maintains container build and runtime definitions, stay with `deploy-container-action`. If you want the fastest path from repository to running service, use `zerodeploy-action`.

> ## Deploy with `mittwald/zerodeploy-action` (Railpack) {#deploy-with-zerodeploy-action-railpack}

Reviewer: Maybe also suggest the `zerodeploy` action as the _first_ alternative? For a more convenience-minded user, it might be useful for this to be at the top of the page, leaving the `deploy-container-action` as an escape hatch for users requiring a higher degree of control

Author: Sure thing. Will apply this combined with restructuring already planned above: Start with easier one, move towards more specific solutions down the line.

The [`mittwald/zerodeploy-action`](https://github.com/mittwald/zerodeploy-action) is designed for repositories that do not have a Docker setup yet. It installs the mittwald CLI, installs Railpack, builds from your repository, and deploys to your target project.

### Required inputs {#required-inputs-zerodeploy}

- **`MITTWALD_API_TOKEN`** secret; API token from mStudio
- **Project ID** (for example `p-XXXXXX`)

### Example workflow {#example-workflow-zerodeploy}

Create `.github/workflows/zerodeploy.yml`:

```yaml
name: Deploy to Mittwald Container Hosting

on:
  workflow_dispatch:

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout repository
        uses: actions/checkout@v4

      - name: Create .env for deployment
        env:
          APP_SECRET: ${{ secrets.APP_SECRET }}
        run: |
          {
            echo "APP_ENV=production"
            echo "APP_SECRET=$APP_SECRET"
          } > .env

      - name: Deploy with zerodeploy

>         uses: mittwald/zerodeploy-action@v1

Reviewer: Caution; for this to work, we need a (moving) `v1` tag in the action repo; currently I see only `v1.1.0` (and older specific tags).

Author: Wait, what? Moving `v1` tag is present and pointing to recent commit. am i missing something?

<img width="732" height="583" alt="image" src="https://github.com/user-attachments/assets/81f0e268-57b3-47c9-9192-038dcf75c41e" />

Reviewer: Huh. I looked into the "versions" field in the marketplace listing, that doesn't list the `v1` tag... 🫠

<img width="300" height="199" alt="Screenshot 2026-05-26 at 16 49 52" src="https://github.com/user-attachments/assets/2d583d1d-d046-4ec2-ab40-0847ac9e1d12" />

OOH. 💡  Because that lists _releases_, and not tags.

Nevermind, then. 😶‍🌫️

        with:
          mittwald-api-token: ${{ secrets.MITTWALD_API_TOKEN }}
          mittwald-project-id: "p-XXXXXX"
```

The action automatically checks for `.env` in `$GITHUB_WORKSPACE/.env` and includes it for deployment when present.

### Useful optional inputs {#useful-optional-inputs-zerodeploy}

- **`repo-subpath`**; deploy only a subdirectory (for monorepos)
- **`node-version`**; Node.js runtime used inside the action (default: `24`)

> - **`mittwald-cli-version`**; CLI version constraint (default: `^1.17.0`)

Reviewer: Tbh. I'm wondering if we should even mention those. These are useful for debugging, but ideally, there should never be any reason for an end user to mess with these settings.

Author: Agreed.

Services not listed are only restarted when the API reports that a restart is actually required.

> ### Next steps {#next-steps}

Reviewer: This could also be a `h2`, I think

```suggestion
## Next steps {#next-steps}
```
